### ISSUE TITLE
DevP2P port uses 30303, ignores provided --port flag

### ISSUE THREAD

**Issue by jakubgs** [erigontech/erigon#8330]
#### System information

Erigon version: `2.49.3`
OS & Version: Ubuntu 22.04
<details><summary>Erigon Command: click to unfold</summary>
<pre>
--chain=holesky
--maxpeers=20
--port=30305
--nat=extip:194.33.40.148
--allow-insecure-unlock
--datadir=/data
--mine=true
--miner.sigfile=/keys/sigfile
--prune=hrtc
--metrics
--metrics.addr=0.0.0.0
--metrics.port=6062
--http
--http.addr=0.0.0.0
--http.port=8547
--http.api=eth,net,erigon,admin,engine
--authrpc.addr=0.0.0.0
--authrpc.port=8553
--authrpc.jwtsecret=/keys/jwtsecret
--v5disc=true
</pre>
</details>
Concensus Layer: Nimbus Beacon Node `23.9.1`
<details><summary>Erigon Command: click to unfold</summary>Concensus Layer Command (with flags/config): click to unfold</summary>
<pre>
--network=holesky
--history=prune
--el=http://localhost:8553
--jwt-secret=jwt.hex
--log-format=json
--log-level=DEBUG
--nat=extip:194.33.40.148
--tcp-port=9002
--udp-port=9002
--max-peers=320
--insecure-netkey-password=true
--sync-light-client=false
--subscribe-all-subnets=false
--doppelganger-detection=true
--num-threads=0
--rest=true
--rest-address=0.0.0.0
--rest-port=9302
--rest-max-body-size=16384
--rest-max-headers-size=128
--metrics=true
--metrics-address=0.0.0.0
--metrics-port=9202
--validator-monitor-auto=true
--validator-monitor-details=false
</pre>
</details>

Chain/Network: `holesky`

#### Expected behaviour
The DevP2P port listens on the port specified via `--port` flag.

#### Actual behaviour
The DevP2P port is listening on `30303` despite flag specifying a different value.
```
/ # ps faux
PID   USER     TIME  COMMAND
    1 root     28:33 /usr/local/bin/erigon --chain=holesky --maxpeers=20 --port=30306 ...

/ # netstat -lpnt | grep -e 3030 -e Proto
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 :::30303                :::*                    LISTEN      1/erigon
tcp        0      0 :::30304                :::*                    LISTEN      1/erigon
```

#### Steps to reproduce the behaviour
Use `--port` with value different than `30303`.

**Comment by jakubgs** [comment]
I just found out about this flag:
```
   --p2p.allowed-ports value [ --p2p.allowed-ports value ]  Allowed ports to pick for different eth p2p protocol versions as follows <porta>,<portb>,..,<porti> (default: 30303, 30304, 30305, 30306, 30307)
```
What kind of silly formatting is this, and what the hell does `portb` or `porti` mean?

**Comment by jakubgs** [comment]
If I specify just one port for `` I get this:
```
EROR Erigon startup    err="run out of allowed ports for p2p eth protocols [40404]. Extend allowed port list via --p2p.allowed-ports"
```
What is this weird behavior? I've never seen this with any other exec layer clients.

**Comment by jakubgs** [comment]
Could someone elucidate the difference between `--port` and `--p2p.allowed-ports`?

**Comment by jakubgs** [comment]
If I use `--port 40403 --p2p.allowed-ports 40404` I get:
```
 > d exec -it test netstat -lpnt
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 0.0.0.0:42069           0.0.0.0:*               LISTEN      1/erigon
tcp        0      0 127.0.0.1:8551          0.0.0.0:*               LISTEN      1/erigon
tcp        0      0 127.0.0.1:8545          0.0.0.0:*               LISTEN      1/erigon
tcp        0      0 127.0.0.1:9090          0.0.0.0:*               LISTEN      1/erigon
tcp        0      0 :::40403                :::*                    LISTEN      1/erigon
tcp        0      0 :::40404                :::*                    LISTEN      1/erigon
```
Why?

**Comment by jakubgs** [comment]
So I'm assuming this is related to this flag:
```
--p2p.protocol value [ --p2p.protocol value ]   Version of eth p2p protocol (default: 68, 67)
```
Which I assume is the number of ports that will be used from the list provided to `--p2p.allowed-ports`.

But why? Why is this something no other client does?

**Comment by jakubgs** [comment]
Is there some kind of magical way to use just the latest version of protocol? Like `--p2p.protocol=latest`?

**Comment by jakubgs** [comment]
It appears the default version is coming from here:
https://github.com/ledgerwatch/erigon/blob/b6ffd87e4480a43a425a727b366cc500c59e5dff/cmd/utils/flags.go#L505-L509
Which in turn comes from here:
https://github.com/ledgerwatch/erigon/blob/b6ffd87e4480a43a425a727b366cc500c59e5dff/node/nodecfg/defaults.go#L52
But I still don't get why there are two protocols enabled on two different ports.

**Comment by jakubgs** [comment]
Apparently - based on suggestions on Discord - the `--port` flag is an "anachronism" and should not be used.

Not sure why it exists then, since it just causes confusion.

**Comment by jakubgs** [comment]
I don't think this is stale at all. The flags are confusing and need cleanup, or help message needs updating.

**Comment by AskAlexSharov** [comment]
right. need or fix it or remove it.

**Comment by lystopad** [comment]
This is fixed on the current 3.x line.

The root cause was that Erigon 2.x ran a **separate `p2p.Server` per eth protocol version**, each binding its own port from the `--p2p.allowed-ports` list (30303, 30304, …). That's why `--port` was effectively ignored and why you saw listeners on 30303/30304 instead of your configured value.

PR #21335 reworked this into a **single shared `p2p.Server`** across all eth protocol versions — one ENR, one Node ID, one listener — and removed the confusing `--p2p.allowed-ports` flag. `--port` is now honored directly.

Verified on v3.5.3 with `--port=30305`:
```
admin_nodeInfo -> "ports": {"discovery": 30305, "listener": 30305}, "listenAddr": "[::]:30305"
log            -> Started P2P networking self=enode://...@...:30305
                  [p2p] GoodPeers eth69=1 eth71=2   (multiple protocols, one port)
```
Both the TCP devp2p listener and the UDP discovery port bind exactly the configured value, and all protocol versions share it.

Closing as fixed — thanks for the detailed report.

**Comment by jakubgs** [comment]
Lovely work, much better. Thanks!
